src/darwin/Framework/CHIPTests/MTRPerControllerStorageTests.m

```diff
@@ -2172,10 +2172,944 @@ - (void)doTestSubscriptionPoolWithSize:(NSInteger)subscriptionPoolSize
+            XCTAssertNotNil(attributePath);
+
+            if ([attributePath.endpoint isEqualToNumber:rootEndpoint] && attributePath.cluster.unsignedLongValue == MTRClusterIDTypeDescriptorID && attributePath.attribute.unsignedLongValue == MTRAttributeIDTypeClusterDescriptorAttributePartsListID) {
+                NSDictionary * data = attributeDict[MTRDataKey];
+                XCTAssertNotNil(data);
+                dataVersionForPartsList = data[MTRDataVersionKey];
+                id dataValue = data[MTRValueKey];
+                XCTAssertNotNil(dataValue);
+                testClusterDataValueForPartsList = [dataValue mutableCopy];
+            }
+        }
+    };
+
+    __block NSMutableArray<NSArray<NSNumber *> *> * initialClusterIndex = [[NSMutableArray alloc] init];
+
+    // TODO: Build the test endpoints from the attribute data value
+    __block NSArray<NSNumber *> * testEndpoints = @[ rootEndpoint, notToBeDeletedEndpoint, toBeDeletedEndpoint ];
+    delegate.onReportEnd = ^{
+        XCTAssertNotNil(dataVersionForPartsList);
+        XCTAssertNotNil(testClusterDataValueForPartsList);
+
+        // Verify that the data store is populated with cluster index and cluster data for endpoints 0, 1 and 2.
+        // Also _persistedClusters and _persistedClusterData has the cluster paths for all clusters in the cluster index.
+        dispatch_sync(self->_storageQueue, ^{
+            XCTAssertTrue([[controller.controllerDataStore _fetchEndpointIndexForNodeID:deviceID] isEqualToArray:testEndpoints]);
+            for (NSNumber * endpoint in testEndpoints) {
+                [initialClusterIndex insertObject:[controller.controllerDataStore _fetchClusterIndexForNodeID:deviceID endpointID:endpoint] atIndex:endpoint.unsignedShortValue];
+                XCTAssertNotNil(initialClusterIndex);
+                for (NSNumber * cluster in [initialClusterIndex objectAtIndex:endpoint.unsignedShortValue]) {
+                    XCTAssertNotNil([controller.controllerDataStore _fetchClusterDataForNodeID:deviceID endpointID:endpoint clusterID:cluster]);
+                    MTRClusterPath * path = [MTRClusterPath clusterPathWithEndpointID:endpoint clusterID:cluster];
+                    XCTAssertTrue([device _persistedClusterContains:path]);
+                    XCTAssertNotNil([device _getPersistedClusterDataForPath:path]);
+                }
+            }
+        });
+        [subscriptionExpectation fulfill];
+    };
+
+    [device setDelegate:delegate queue:queue];
+
+    [self waitForExpectations:@[ subscriptionExpectation ] timeout:60];
+
+    // Inject a fake attribute report deleting endpoint 2 from the parts list at the root endpoint.
+    dataVersionForPartsList = [NSNumber numberWithUnsignedLongLong:(dataVersionForPartsList.unsignedLongLongValue + 1)];
+    // TODO: try to delete entries from the actual data received in the initila attribute report to make the neunsignedIntegerArrayValue
```

Comment:
I have no idea what `neunsignedIntegerArrayValue` is.

Also, why _aren't_ we just deleting entries from the actual data received?

This stuff is all super-fragile and will break if things changed about the all-clusters app.  At least we need issues tracking these TODOs.

Reply:
done. we are deleting the entries from actual data.